crypto/err/err_prn.c

```diff
@@ -44,6 +45,115 @@ void ERR_print_errors_cb(int (*cb) (const char *str, size_t len, void *u),
+    const char *data = NULL;
+    int flags;
+    unsigned long err = ERR_peek_last_error();
+
+    if (separator == NULL)
+        separator = "";
+    if (err == 0)
+        put_error(ERR_LIB_CMP, NULL, 0, "", 0);
+
+    do {
+        size_t available_len, data_len;
+        const char *curr = txt, *next = txt;
+        const char *leading_separator = separator;
+        int trailing_separator = 0;
+        char *tmp;
+
+        ERR_peek_last_error_all(&file, &line, &func, &data, &flags);
+        if ((flags & ERR_TXT_STRING) == 0) {
+            data = "";
+            leading_separator = "";
+        }
+        data_len = strlen(data);
+
+        /* workaround for limit of ERR_print_errors_cb() */
+        if (data_len >= MAX_DATA_LEN
+                || strlen(separator) >= (size_t)(MAX_DATA_LEN - data_len))
+            available_len = 0;
+        else
+            available_len = MAX_DATA_LEN - data_len - strlen(separator) - 1;
+        /* MAX_DATA_LEN > available_len >= 0 */
+
+        if (*separator == '\0') {
+            const size_t len_next = strlen(next);
+
+            if (len_next <= available_len) {
+                next += len_next;
+                curr = NULL; /* no need to split */
+            } else {
+                next += available_len;
+                curr = next; /* will split at this point */
+            }
+        } else {
+            while (*next != '\0' && (size_t)(next - txt) <= available_len) {
+                curr = next;
+                next = strstr(curr, separator);
+                if (next != NULL) {
+                    next += strlen(separator);
+                    trailing_separator = *next == '\0';
+                } else {
+                    next = curr + strlen(curr);
+                }
+            }
+            if ((size_t)(next - txt) <= available_len)
+                curr = NULL; /* the above loop implies *next == '\0' */
+        }
+        if (curr != NULL) {
+            /* split error msg at curr since error data would get too long */
+            if (curr != txt) {
+                tmp = OPENSSL_strndup(txt, curr - txt);
+                if (tmp == NULL)
+                    return;
+                ERR_add_error_data(2, separator, tmp);
+                OPENSSL_free(tmp);
+            }
+            put_error(ERR_LIB_CMP, func, err, file, line);
+            txt = curr;
+        } else {
+            if (trailing_separator) {
+                tmp = OPENSSL_strndup(txt, next - strlen(separator) - txt);
+                if (tmp == NULL)
+                    return;
+                /* output txt without the trailing separator */
+                ERR_add_error_data(2, leading_separator, tmp);
+                OPENSSL_free(tmp);
+            } else {
+                ERR_add_error_data(2, leading_separator, txt);
+            }
+            txt = next; /* finished */
+        }
+    } while (*txt != '\0');
+}
+
+void ERR_add_error_mem_bio(const char *separator, BIO *bio)
```

Comment:
Actually I am not very convinced that printing text of arbitrary size into a memory bio
and pushing that afterwards to the error queue is a good idea at all.
The error queue will always be limited in size, therefore no guarantee that all the text survives.
I would let the user pass in a BIO where a transcript of the whole process is collected,
and let the user pass it in as a parameter, so it can be printed to the terminal, or discarded
if all works fine, if that is a memory BIO or a file BIO could be the choice of the user.

Reply:
Well, you are right that also the solution implemented here has its limits (due to the limit of the number of error queue entries, which BTW could be relaxed to some extent at least).

Yet for the intended use case the given implementation is (in all likelihood) sufficient, namely:
In case a certificate verification failed, the new cert verify callback function `X509_STORE_CTX_print_verify_cb()` prints relevant portions not only of the failed cert itself, but - where it is relevant - also the list of untrusted certs and trusted certs contained in the `X509_STORE_CTX`. 
It turned out that occasionally this pretty helpful information is truncated by the 4096 bytes error data limit per error queue entry. With the help of `ERR_add_error_txt()` the cert info output can overflow, if needed, into a second error queue entry, and more than 2*4096 characters are not (reasonably) needed for this use case.

Letting the caller/user optionally pass in a BIO for receiving extended diagnostics would certainly be a more flexible solution, but at the expense of further complication of the cert verify API, 
and I fear that the typical programmer won't spend the effort to provide and handle such a BIO, such that in effect the end user will likely not have the benefit of the extended diagnostic info provided by `X509_STORE_CTX_print_verify_cb()`.

Reply:
Thanks to @DDvO, I would like to resolve this comment. @bernd-edlinger is it OK for your?